clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java

```diff
@@ -109,101 +134,102 @@ public void run() {
-                this.exception.set(Optional.of(new RuntimeException(t)));
+            log.error("The background thread failed due to unexpected error", t);
+            throw new RuntimeException(t);
         } finally {
             close();
             log.debug("{} closed", getClass());
         }
     }
 
     /**
-     * Process event from a single poll
+     * Process event from a single poll. It performs the following tasks sequentially:
+     *  1. Try to poll and event from the queue, and try to process it.
+     *  2. Try to find Coordinator if needed
+     *  3. Try to send fetches.
+     *  4. Poll the networkClient for outstanding requests. If non: poll until next
+     *  iteration.
      */
     void runOnce() {
-        this.inflightEvent = maybePollEvent();
-        if (this.inflightEvent.isPresent()) {
-            log.debug("processing application event: {}", this.inflightEvent);
+        Optional<ApplicationEvent> event = maybePollEvent();
+
+        if (event.isPresent()) {
+            log.debug("processing application event: {}", event);
+            consumeApplicationEvent(event.get());
         }
-        if (this.inflightEvent.isPresent() && maybeConsumeInflightEvent(this.inflightEvent.get())) {
-            // clear inflight event upon successful consumption
-            this.inflightEvent = Optional.empty();
+
+        final long currentTimeMs = time.milliseconds();
+        long pollWaitTimeMs = timeToNextHeartbeatMs();
+
+        // TODO: Add a condition here, like shouldFindCoordinator in the future.  Since we don't always need to find
+        //  the coordinator.
+        if (coordinatorManager.isPresent()) {
+            pollWaitTimeMs = Math.min(pollWaitTimeMs,
+                    handlePollResult(coordinatorManager.get().poll(currentTimeMs)));
         }
 
         // if there are pending events to process, poll then continue without
         // blocking.
-        if (!applicationEventQueue.isEmpty() || inflightEvent.isPresent()) {
-            networkClient.poll(time.timer(0));
+        if (!applicationEventQueue.isEmpty()) {
+            networkClientDelegate.poll(time.timer(0), false);
             return;
         }
         // if there are no events to process, poll until timeout. The timeout
         // will be the minimum of the requestTimeoutMs, nextHeartBeatMs, and
         // nextMetadataUpdate. See NetworkClient.poll impl.
-        networkClient.poll(time.timer(timeToNextHeartbeatMs(time.milliseconds())));
+        networkClientDelegate.poll(time.timer(pollWaitTimeMs), false);
+    }
+
+    long handlePollResult(NetworkClientDelegate.PollResult res) {
+        Objects.requireNonNull(res);
+        if (!res.unsentRequests.isEmpty()) {
+            networkClientDelegate.addAll(res.unsentRequests);
+            return Long.MAX_VALUE;
+        }
+        return res.timeMsTillNextPoll;
     }
 
-    private long timeToNextHeartbeatMs(final long nowMs) {
+    private long timeToNextHeartbeatMs() {
```

Comment:
Let's get rid of this for now. I imagine we'll have a HeartbeatManager or something like that later.